### ISSUE TITLE
 Cannot cast types (BYTE, ) error cpc/msx1 target  (regression bug)

### ISSUE THREAD

**Issue by ldir-hector** [spotlessmind1975/ugbasic#562]
The following code sample : 
```
BITMAP ENABLE (160,200,16)
CLS
DIM monCosinus AS SIGNED 32BIT(180) = #{9998,9994,9986,9976,9962,9945,9925,9903,9877,9848,9816,9781,9744,9703,9659,9613,9563,9511,9455,9397,9336,9272,9205,9135,9063,8988,8910,8829,8746,8660,8572,8480,8387,8290,8192,8090,7986,7880,7771,7660,7547,7431,7314,7193,7071,6947,6820,6691,6561,6428,6293,6157,6018,5878,5736,5592,5446,5299,5150,5000,4848,4695,4540,4384,4226,4067,3907,3746,3584,3420,3256,3090,2924,2756,2588,2419,2250,2079,1908,1736,1564,1392,1219,1045,872,698,523,349,175,0,-175,-349,-523,-698,-872,-1045,-1219,-1392,-1564,-1736,-1908,-2079,-2250,-2419,-2588,-2756,-2924,-3090,-3256,-3420,-3584,-3746,-3907,-4067,-4226,-4384,-4540,-4695,-4848,-5000,-5150,-5299,-5446,-5592,-5736,-5878,-6018,-6157,-6293,-6428,-6561,-6691,-6820,-6947,-7071,-7193,-7314,-7431,-7547,-7660,-7771,-7880,-7986,-8090,-8192,-8290,-8387,-8480,-8572,-8660,-8746,-8829,-8910,-8988,-9063,-9135,-9205,-9272,-9336,-9397,-9455,-9511,-9563,-9613,-9659,-9703,-9744,-9781,-9816,-9848,-9877,-9903,-9925,-9945,-9962,-9976,-9986,-9994,-9998,-10000}
dec_x = (SIGNED 32BIT) 0
PROCEDURE tir
	
	dec_x =  monCosinus(angle) / 2
END PROC
```
was compiling on ugBasic 1.12, now I got the error message Cannot cast types (BYTE, ) for the line dec_x =  monCosinus(angle) / 2

That's quite surprising because if I remove the division, it works and if I do such a division on a simple value like 200 for instance, it works too.

**Comment by spotlessmind1975** [comment]
Hello @ldir-hector , and **thank you for your bug report!**

Actually, it's not a regression but a wrong behavior of version 1.12 (and earlier, I think up to 1.11 or something like that, I haven't looked into). 

In a nutshell, ugBASIC ensures that variables defined within a procedure are accessible only within the procedure itself. It is the so-called concept of "local variable" as opposed to the concept of "global variable". In the example source, the `monCosinus` array is defined in the main program and then accessed within the procedure. However, for ugBASIC they are two different arrays. In particular, the array accessed within the procedure will be a different array than the global one. However (and this is the bug that will be fixed in the next coldfix), the local array is not "typed" so to speak. As a result, when trying to assign the value fetched from the array to the `dec_x` variable, ugBASIC is unable to deduce what type it is -- hence the error.

You don't need to wait for the coldfix anyway, also because it would actually hide the real error instead of making it clear. There are two solutions, than can be applied at once.

The first is to define `dec_x` and `monCosinus` variables as "global" outside the procedure:

    GLOBAL dec_x
    GLOBAL monCosinus
    PROCEDURE tir
    ....

b) define the `dec_x` and `monCosinus` variables as "shared" with the main program:

    PROCEDURE tir
       SHARED dec_x
       SHARED monCosinus
       ....

Finally, using the `OPTION EXPLICIT` command would have made the problem clearer.

**Comment by spotlessmind1975** [comment]
Fixed with COLDFIX v1.13 [rev. 20230212.0014].
